File: algorithms/81. Search in Rotated Sorted Array II/solution.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        low, high = 0, len(nums) - 1
        while low <= high:
            mid = low + (high - low) / 2
            if nums[mid] == target:
                return True
            # left is sorted
            if nums[mid] > nums[high]:
                if nums[low] <= target < nums[mid]:
                    high = mid - 1
                else:
                    low = mid + 1
            # right is sorted
            elif nums[mid] < nums[high]:
                if nums[mid] < target <= nums[high]:
                    low = mid + 1
                else:
                    high = mid - 1
            # skip dup
            else:
                high -= 1
        return False

File: algorithms/81. Search in Rotated Sorted Array II/solution.cc
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int low = 0, high = nums.size() - 1, mid;
        while ( low <= high ) {
            mid = low + (high - low) / 2;
            if ( nums[mid] == target ) {
                return true;
            }
            if ( nums[mid] > nums[high] ) {
                if ( nums[low] <= target && target < nums[mid] ) {
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            } else if ( nums[mid] < nums[high] ) {
                if ( nums[mid] < target && target <= nums[high] ) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            } else {
                --high;
            }
        }
        return false;
    }
};
